### DISCUSSION TITLE
Dependencies Conflict with Airflow providers google - (Google Ads, GCS Sensor)

### DISCUSSION THREAD

**Discussion by igbiorczyk** [apache/airflow#24766] [General]
### Apache Airflow version

2.3.0

### What happened

Hi,

I'm trying to get Google-ads data from the Google ads API client library python (https://github.com/googleads/google-ads-python).
Everything is okay when starting the function manually. While trying to put it in a DAG, there is a dependencies conflict inside Airflow.
I'm just importing the google ads lib in the DAG file. The idea was to just use a pythonOperator.

**from google.ads.googleads.client import GoogleAdsClient**

Broken DAG: [/data_eng/airflow/dags/google-ads_to_bigquery.py] Traceback (most recent call last): File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 886, in require needed = self.resolve(parse_requirements(requirements)) File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 777, in resolve raise VersionConflict(dist, req).with_context(dependent_req) pkg_resources.ContextualVersionConflict: (proto-plus 1.20.3 (/usr/local/lib/python3.9/dist-packages), Requirement.parse('proto-plus==1.19.6'), {'google-ads'})

Weirdly  i can't find any lib requiring proto-plus 1.20.3 in my python 3.9 distribution. Moving back to proto-plus doesn't resolve issue and start a whole serie of lib issue, which mean to downgrade a lot of them (google-api-core for example)

I then tryed to use the Airflow Provider Google Ads to GCS (https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/ads.html). Using the following import : 
**from airflow.providers.google.ads.transfers.ads_to_gcs import GoogleAdsToGcsOperator**

But i got the same issue as from the API.
![image](https://user-images.githubusercontent.com/90033810/176688068-e5e6d550-5b59-4327-9d33-8500667818e6.png)

I find that a similar issue is reported and Airflow team is working on a fix for the google ads lib but i don't understand why the provider return the same issue.

While trying to use the Airflow Google Provider for others actions (gcs sensor) i still got a lot of dependencies issue even if my lib is updated. For example with this import 

**from airflow.providers.google.cloud.sensors.gcs import GCSObjectExistenceSensor**
![image](https://user-images.githubusercontent.com/90033810/176689182-03d371a2-5cc5-4fc5-8f94-15ff4aa54383.png)
And i don't understand why it's looking for the precise google-api-core 2.7.3 despite beeing on 2.8.2

I'm not sure if all the issues are related and i'm sorry if not but doesn't anyone has an idea related ?
Have a nice day !

### What you think should happen instead

Well.... working import

### How to reproduce

One of the following for each error
**from google.ads.googleads.client import GoogleAdsClient**
**from airflow.providers.google.ads.transfers.ads_to_gcs import GoogleAdsToGcsOperator**
**from airflow.providers.google.cloud.sensors.gcs import GCSObjectExistenceSensor**
### Operating System

Debian 11

### Versions of Apache Airflow Providers

I'm using apache-airflow-providers-google    8.1.0

### Deployment

Other

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
You have not described how you installed airflow and providers. Could you please do so? Did you use constraints ? What was your sequence of upgrading the providers? You seem to not use the provider that comes with Airflow.

Can you also provide a dump of all deps you have in your venv where you have airflow installed: `pip freeze` will provide it.

Python dependency system is complex and a lot of dependencies does not specify the **right** set of required dependencies which migh lead to unexpected situations. It might be that simply ads functionality is broken on airflow 2.3.0 + google-providedr 8.1.0 combo - because for whatever reason they have not specified the proto-plus depenendency in the setup.py. It's simply impossible to check all the combinations and all the featurs of various versions of the providers with various versions of Airlfow, so things like that **might** happen.

If my guess is right - simply google provider 8.1.0 and Airflow 2.3.0 will simply never get a working "ads" becuase of the conflicts in requirements (which is not even visible in the package requirements so you can happily install them together). There is not much we can do if this is the case (except dumping min-version of airlfow in google provider 8.1.1 ). You are a bit on your own there to go back with versions of provider and finding out which one will work for you (7.0.0 maybe?) or 6.*.. It does not invalidate whole "google" provider,  just the ads might not work in this case (BTW. In the future - probably near - we will split-off ads from google provider most likely - we make a split test right now in #23747 and when we get it right, we might also do similar split for Google one.

However not all is lost for you in this case. Proto-plus seems to have been updated in 2.3.1 and 2.3.2 - and there is a high chance, the ads feature will work there - and you should do it anyway, because 2.3.1 and 2.3.2 have  important bug-fixes that you should for sure include. 

So I suggest the following course of actions:

1) explain how you installed your airlfow and providers (I might then be able to help better)
2) perform migration of airflow to 2.3.2 and (possibly after you tell me how you upgraded and I confirm this is fine)

I hope this will help.

**Reply by potiuk** [reply]
BTW. In a few days ~ week we are going to have Airlfow 2.3.3 release which will be linked to the latest 8.1.0 provider so it will likely **just work** for you. Still upgrading to 2.3.2 as an interim solution might at least verify the hypothesis of mine.

**Reply by igbiorczyk** [reply]
Sorry for the long delay of answer @potiuk 

We installed Airflow with the PyPi install instructions from official documentation : https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

We strictly use Pip to install providers, lib and upgrading it.

I will start to work on migration to airflow 2.3.2 asap

Here is the list of deb we have currently : 
aiohttp==3.8.1
aiosignal==1.2.0
alembic==1.7.7
anyio==3.5.0
apache-airflow==2.3.0
apache-airflow-providers-ftp==2.1.2
apache-airflow-providers-google==8.1.0
apache-airflow-providers-http==2.1.2
apache-airflow-providers-imap==2.2.3
apache-airflow-providers-postgres==4.1.0
apache-airflow-providers-sqlite==2.1.3
apispec==3.3.2
argcomplete==1.12.3
async-generator==1.10
async-timeout==4.0.2
attrs==20.3.0
Babel==2.10.1
bcrypt==3.2.2
billiard==4.0.0
blinker==1.4
boto3==1.23.1
botocore==1.26.1
cached-property==1.5.2
cachelib==0.6.0
cachetools==4.2.4
cattrs==1.10.0
certifi==2021.10.8
cffi==1.15.0
charset-normalizer==2.0.12
click==7.1.2
clickclick==20.10.2
colorama==0.4.4
colorlog==4.8.0
commonmark==0.9.1
connexion==2.13.1
cron-descriptor==1.2.24
croniter==1.3.4
cryptography==37.0.2
cycler==0.11.0
db-dtypes==1.0.1
dbus-python==1.2.16
defusedxml==0.7.1
Deprecated==1.2.13
dill==0.3.4
distlib==0.3.4
distro-info==1.0
dnspython==2.2.1
docutils==0.18.1
elastic-transport==8.1.2
elasticsearch==7.17.3
elasticsearch-curator==5.8.4
email-validator==1.2.1
et-xmlfile==1.1.0
filelock==3.7.1
Flask==1.1.4
Flask-AppBuilder==3.4.5
Flask-Babel==2.0.0
Flask-Caching==1.10.1
Flask-JWT-Extended==3.25.1
Flask-Login==0.4.1
Flask-OpenID==1.3.0
Flask-Session==0.4.0
Flask-SQLAlchemy==2.5.1
Flask-WTF==0.14.3
fonttools==4.33.3
frozenlist==1.3.0
future==0.18.2
geoip2==4.5.0
google-ads==16.0.0
google-api-core==2.8.2
google-api-python-client==1.12.11
google-auth==1.35.0
google-auth-httplib2==0.1.0
google-auth-oauthlib==0.5.1
google-cloud-aiplatform==1.14.0
google-cloud-appengine-logging==1.1.2
google-cloud-audit-log==0.2.2
google-cloud-automl==2.7.3
google-cloud-bigquery==2.34.4
google-cloud-bigquery-datatransfer==3.6.2
google-cloud-bigquery-storage==2.13.1
google-cloud-bigtable==1.7.2
google-cloud-build==3.8.3
google-cloud-container==2.10.8
google-cloud-core==2.3.0
google-cloud-datacatalog==3.8.1
google-cloud-dataplex==1.0.1
google-cloud-dataproc==4.0.3
google-cloud-dataproc-metastore==1.5.1
google-cloud-dlp==1.0.2
google-cloud-kms==2.11.2
google-cloud-language==1.3.2
google-cloud-logging==3.1.2
google-cloud-memcache==1.3.2
google-cloud-monitoring==2.9.2
google-cloud-orchestration-airflow==1.3.2
google-cloud-os-login==2.6.2
google-cloud-pubsub==2.13.0
google-cloud-redis==2.8.1
google-cloud-resource-manager==1.5.1
google-cloud-secret-manager==1.0.2
google-cloud-spanner==1.19.3
google-cloud-speech==1.3.4
google-cloud-storage==1.44.0
google-cloud-tasks==2.9.1
google-cloud-texttospeech==1.0.3
google-cloud-translate==1.7.2
google-cloud-videointelligence==1.16.3
google-cloud-vision==1.0.2
google-cloud-workflows==1.6.3
google-crc32c==1.3.0
google-resumable-media==2.3.2
googleapis-common-protos==1.56.3
graphviz==0.20
greenlet==1.1.2
grpc-google-iam-v1==0.12.4
grpcio==1.46.0
grpcio-gcp==0.2.2
grpcio-status==1.46.0
gspread==5.3.2
gunicorn==20.1.0
h11==0.12.0
httpcore==0.14.7
httplib2==0.20.4
httpx==0.22.0
idna==3.3
importlib-metadata==4.11.3
inflection==0.5.1
isodate==0.6.1
itsdangerous==1.1.0
Jinja2==2.11.3
jmespath==1.0.0
joblib==1.1.0
json-merge-patch==0.2
jsonschema==4.5.1
kiwisolver==1.4.2
lazy-object-proxy==1.7.1
lockfile==0.12.2
looker-sdk==22.4.0
lxml==4.8.0
Mako==1.2.0
Markdown==3.3.7
MarkupSafe==2.0.1
marshmallow==3.15.0
marshmallow-enum==1.5.1
marshmallow-oneofschema==3.0.1
marshmallow-sqlalchemy==0.26.1
matplotlib==3.5.2
maxminddb==2.2.0
multidict==6.0.2
mysql-connector==2.2.9
nltk==3.7
nox==2020.12.31
numpy==1.22.3
oauth2client==4.1.3
oauthlib==3.2.0
openpyxl==3.0.10
outcome==1.1.0
packaging==21.3
pandas==1.4.2
pandas-gbq==0.17.6
paramiko==2.11.0
pathspec==0.9.0
pendulum==2.1.2
Pillow==9.1.0
platformdirs==2.5.2
pluggy==1.0.0
prison==0.2.1
proto-plus==1.19.6
protobuf==3.20.0
psutil==5.9.0
psycopg2-binary==2.9.3
py==1.11.0
pyarrow==6.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pydata-google-auth==1.4.0
Pygments==2.12.0
PyJWT==1.7.1
PyNaCl==1.5.0
pyOpenSSL==22.0.0
pyparsing==3.0.8
pyrsistent==0.18.1
pysftp==0.2.9
PySocks==1.7.1
python-apt==2.2.1
python-daemon==2.3.0
python-dateutil==2.8.2
python-nvd3==0.15.0
python-slugify==6.1.2
python3-openid==3.2.0
pytz==2022.1
pytzdata==2020.1
PyYAML==5.4.1
regex==2022.4.24
requests==2.27.1
requests-aws4auth==1.1.2
requests-file==1.5.1
requests-oauthlib==1.3.1
requests-toolbelt==0.9.1
rfc3986==1.5.0
rich==12.4.1
rsa==4.8
s3transfer==0.5.2
scikit-learn==1.1.0
scipy==1.8.0
scp==0.14.4
selenium==4.1.5
setproctitle==1.2.3
six==1.16.0
sniffio==1.2.0
sortedcontainers==2.4.0
SQLAlchemy==1.4.9
sqlalchemy-bigquery==1.4.4
SQLAlchemy-JSONField==1.0.0
SQLAlchemy-Utils==0.38.2
stop-words==2018.7.23
supervisor==4.2.4
swagger-ui-bundle==0.0.9
tableauserverclient==0.18.0
tabulate==0.8.9
tenacity==8.0.1
termcolor==1.1.0
text-unidecode==1.3
threadpoolctl==3.1.0
tqdm==4.64.0
trio==0.20.0
trio-websocket==0.9.2
typing-extensions==4.2.0
unattended-upgrades==0.1
unicodecsv==0.14.1
Unidecode==1.3.4
uritemplate==3.0.1
urllib3==1.26.4
virtualenv==20.15.1
voluptuous==0.13.1
webdavclient3==3.14.6
Werkzeug==1.0.1
wordcloud==1.8.1
wrapt==1.14.1
wsproto==1.1.0
WTForms==2.3.3
xmltodict==0.13.0
yarl==1.7.2
zeep==4.1.0
zenpy==2.0.24
zipp==3.8.0

**Reply by potiuk** [reply]
I suggest - just go straight to 2.3.3 or wait a few days - we are working on 2.3.4 release which will include even newer set of dependencies and providers.

And yeah, jus to explain -  there might be "partial" incompatibilities of some features for some combinations of providers + airflow - like in this case - newer version of provider x might conflict when it comes to some dependencies. This is inevitable and we cannot do much about it, but upgrading to latest version of Airflow in the line you are using should usually be painless (it only contains bugfixes and dependency updates) and is the best way to solve those kinds of conflicts.
